libvast/src/system/table_index.cpp

```diff
@@ -0,0 +1,296 @@
+          detail::overload(
+            [&](const attribute_extractor& ex, const data& x) {
+              return lookup_impl(p, ex, x);
+            },
+            [&](const data_extractor& dx, const data& x) {
+              return lookup_impl(p, dx, x);
+            },
+            [&](const auto&, const auto&) -> expected<bitmap> {
+              // Ignore unexpected lhs/rhs combinations.
+              return bitmap{};
+            }
+          ),
+          p.lhs, p.rhs);
+      },
+      [&](const none&) -> expected<bitmap> {
+        return bitmap{};
+      }),
+    expr);
+}
+
+caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
+                                               const attribute_extractor& ex,
+                                               const data& x) {
+  VAST_TRACE(VAST_ARG(pred), VAST_ARG(ex), VAST_ARG(x));
+  VAST_IGNORE_UNUSED(x);
+  // We know that the columns vector contains two meta fields: time at index
+  // 0 and type at index 1.
+  static_assert(table_index::meta_column_count == 2);
+  VAST_ASSERT(columns_.size() >= table_index::meta_column_count);
+  if (ex.attr == "time") {
+    VAST_ASSERT(is<timestamp>(x));
+    auto fac = [&] { return make_time_index(meta_dir() / "time"); };
+    return with_meta_column(0, fac, [&](column_index& col) {
+      return col.lookup(pred);
+    });
+  } else if (ex.attr == "type") {
+    VAST_ASSERT(is<std::string>(x));
+    auto fac = [&] { return make_type_index(meta_dir() / "type"); };
+    return with_meta_column(1, fac, [&](column_index& col) {
+      return col.lookup(pred);
+    });
+  }
+  VAST_WARNING("unsupported attribute:", ex.attr);
+  return ec::invalid_query;
+}
+
+caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
+                                               const data_extractor& dx,
+                                               const data& x) {
+  VAST_TRACE(VAST_ARG(pred), VAST_ARG(dx), VAST_ARG(x));
+  VAST_IGNORE_UNUSED(x);
+  if (dx.offset.empty()) {
+    VAST_ASSERT(num_data_columns() == 1);
+    auto fac = [&] { return make_flat_data_index(data_dir(), event_type_); };
+    return with_data_column(0, fac, [&](column_index& col) {
+      return col.lookup(pred);
+    });
+  } else {
+    auto r = get<record_type>(dx.type);
+    auto k = r.resolve(dx.offset);
+    VAST_ASSERT(k);
+    auto t = r.at(dx.offset);
+    VAST_ASSERT(t);
+    auto fac = [&] {
+      auto p = data_dir();
+      for (auto& x : *k)
+        p /= x;
+      return make_field_data_index(p, *t, dx.offset);
+    };
+    auto index = r.flat_index_at(dx.offset);
+    if (!index) {
+      VAST_DEBUG("invalid offset for record type", dx.type);
+      return bitmap{};
+    }
+    return with_data_column(*index, fac, [&](column_index& col) {
+      return col.lookup(pred);
+    });
+  }
+  return bitmap{};
+}
+
+// -- constructors, destructors, and assignment operators ----------------------
+
+table_index::table_index(type event_type, path base_dir)
```

Comment:
Let's rename `event_type` to `layout` everywhere in this PR, since the notion of event no longer really exists at this point.